From 900f9dc0484c000e2054d447996b4c57929ea05a Mon Sep 17 00:00:00 2001
From: "pj.chen" <chen.pj@inventec.com>
Date: Thu, 3 Mar 2022 10:26:28 +0000
Subject: [PATCH 05/31] App - Compose Aux Firmware Rev Info in Get Device Id
 command

Symptom/Reason :
    - Aux Firmware Rev Info in "Get Device Id" command response is null.

Root Cause:
    - The feature is not implemented

Solution/Change:
    [os-release]
        - Redefine VERSION_ID field format
    [phosphor-ipmi-host]
        - Apply new version id parsing rule for Inventec

Entry Test:
ipmitool mc info
---
 apphandler.cpp | 48 +++++++++++++++++++++++-------------------------
 1 file changed, 23 insertions(+), 25 deletions(-)

diff --git a/apphandler.cpp b/apphandler.cpp
index 6cdd78f..71b291e 100644
--- a/apphandler.cpp
+++ b/apphandler.cpp
@@ -498,7 +498,7 @@ int convertVersion(std::string s, Revision& rev)
         if (location != std::string::npos)
         {
             rev.major =
-                static_cast<char>(std::stoi(s.substr(0, location), 0, 10));
+                static_cast<char>(std::stoi(s.substr(0, location), 0));
             token = s.substr(location + 1);
         }
 
@@ -508,7 +508,7 @@ int convertVersion(std::string s, Revision& rev)
             if (location != std::string::npos)
             {
                 rev.minor = static_cast<char>(
-                    std::stoi(token.substr(0, location), 0, 10));
+                    std::stoi(token.substr(0, location), 0));
                 token = token.substr(location + 1);
             }
         }
@@ -516,39 +516,37 @@ int convertVersion(std::string s, Revision& rev)
         // Capture the number of commits on top of the minor tag.
         // I'm using BE format like the ipmi spec asked for
         location = token.find_first_of(".-");
+        if (location != std::string::npos)
+        {
+            token = token.substr(location + 1);
+        }
         if (!token.empty())
         {
-            commits = std::stoi(token.substr(0, location), 0, 16);
-            rev.d[0] = (commits >> 8) | (commits << 8);
-
-            // commit number we skip
-            location = token.find_first_of(".-");
-            if (location != std::string::npos)
-            {
-                token = token.substr(location + 1);
-            }
+            // Extract 1st year field
+            commits = std::stoi(token.substr(0, 2));
+            rev.d[0] = commits;
+            // Extract 2nd year field
+            commits = std::stoi(token.substr(2, 2));
+            rev.d[0] |= commits << 8;
+
+            // Extract month field
+            commits = std::stoi(token.substr(4, 2));
+            rev.d[1] = commits;
         }
         else
         {
             rev.d[0] = 0;
         }
 
+        // Extract RC revision value
+        location = token.find("-rc");
         if (location != std::string::npos)
         {
-            token = token.substr(location + 1);
-        }
-
-        // Any value of the optional parameter forces it to 1
-        location = token.find_first_of(".-");
-        if (location != std::string::npos)
-        {
-            token = token.substr(location + 1);
+            token = token.substr(location + 3);
+            location = token.find_first_of(".-");
+            commits = std::stoi(token.substr(0, location));
+            rev.d[1] |= commits << 8;
         }
-        commits = (!token.empty()) ? 1 : 0;
-
-        // We do this operation to get this displayed in least significant bytes
-        // of ipmitool device id command.
-        rev.d[1] = (commits >> 8) | (commits << 8);
     }
 
     return 0;
@@ -643,7 +641,7 @@ ipmi::RspType<uint8_t,  // Device ID
                 devId.addnDevSupport = data.value("addn_dev_support", 0);
                 devId.manufId = data.value("manuf_id", 0);
                 devId.prodId = data.value("prod_id", 0);
-                devId.aux = data.value("aux", 0);
+                devId.aux = rev.d[1] << 16 | rev.d[0];
 
                 // Set the availablitity of the BMC.
                 defaultActivationSetting = data.value("availability", true);
-- 
2.33.0

